src/coprocessor/dag/aggr_fn/impl_first.rs

```diff
@@ -0,0 +1,249 @@
+
+use std::marker::PhantomData;
+
+use cop_codegen::AggrFunction;
+use cop_datatype::EvalType;
+use tipb::expression::{Expr, ExprType, FieldType};
+
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::Tz;
+use crate::coprocessor::dag::expr::EvalContext;
+use crate::coprocessor::dag::rpn_expr::{RpnExpression, RpnExpressionBuilder};
+use crate::coprocessor::Result;
+
+/// The parser for FIRST aggregate function.
+pub struct AggrFnDefinitionParserFirst;
+
+impl super::AggrDefinitionParser for AggrFnDefinitionParserFirst {
+    fn check_supported(&self, aggr_def: &Expr) -> Result<()> {
+        assert_eq!(aggr_def.get_tp(), ExprType::First);
+        if aggr_def.get_children().len() != 1 {
+            return Err(box_err!(
+                "Expect 1 parameter, but got {}",
+                aggr_def.get_children().len()
+            ));
+        }
+
+        // Check whether parameter expression is supported.
+        RpnExpressionBuilder::check_expr_tree_supported(&aggr_def.get_children()[0])?;
+
+        Ok(())
+    }
+
+    fn parse(
+        &self,
+        mut aggr_def: Expr,
+        time_zone: &Tz,
+        max_columns: usize,
+        out_schema: &mut Vec<FieldType>,
+        out_exp: &mut Vec<RpnExpression>,
+    ) -> Result<Box<dyn super::AggrFunction>> {
+        use cop_datatype::FieldTypeAccessor;
+        use std::convert::TryFrom;
+        assert_eq!(aggr_def.get_tp(), ExprType::First);
+        let child = aggr_def.take_children().into_iter().next().unwrap();
+        let eval_type = EvalType::try_from(child.get_field_type().tp()).unwrap();
+
+        // FIRST outputs one column with the same type as its child
+        out_schema.push(aggr_def.take_field_type());
+
+        // FIRST doesn't need to cast, so using the expression directly.
+        out_exp.push(RpnExpressionBuilder::build_from_expr_tree(
+            child,
+            time_zone,
+            max_columns,
+        )?);
+
+        match_template_evaluable! {
+            TT, match eval_type {
+                EvalType::TT => Ok(Box::new(AggrFnFirst::<TT>::new()))
+            }
+        }
+    }
+}
+
+/// The FIRST aggregate function.
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggrFnStateFirst::<T>::new())]
+pub struct AggrFnFirst<T>(PhantomData<T>)
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>;
+
+impl<T> AggrFnFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    fn new() -> Self {
+        AggrFnFirst(PhantomData)
+    }
+}
+
+/// The state of the FIRST aggregate function.
+#[derive(Debug)]
+pub enum AggrFnStateFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    Empty,
+    Valued(Option<T>),
+}
+
+impl<T> AggrFnStateFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    pub fn new() -> Self {
+        AggrFnStateFirst::Empty
+    }
+}
+
+// default impl all `Evaluable` for all `AggrFnStateFirst` to make `AggrFnStateFirst`
```

Comment:
Maybe you need to explain why we choose to manually implement `AggrFunctionState` instead of `ConcreteAggrFunctionState` in the comment.